subxt/src/backend/unstable/follow_stream_unpin.rs

```diff
@@ -0,0 +1,634 @@
+
+/// The future returned from [`UnpinMethod`].
+pub type UnpinFut = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+impl<Hash: BlockHash> std::marker::Unpin for FollowStreamUnpin<Hash> {}
+
+impl<Hash: BlockHash> Stream for FollowStreamUnpin<Hash> {
+    type Item = Result<FollowStreamMsg<BlockRef<Hash>>, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut();
+
+        loop {
+            // Poll the unpin tasks while they are completing. if we get back None, then
+            // no tasks in the list, and if pending, we'll be woken when we can poll again.
+            if let Poll::Ready(Some(())) = this.unpin_futs.poll_next_unpin(cx) {
+                continue;
+            };
+
+            // Poll the inner stream for the next event.
+            let Poll::Ready(ev) = this.inner.poll_next_unpin(cx) else {
+                return Poll::Pending;
+            };
+
+            // No more progress to be made if inner stream done.
+            let Some(ev) = ev else {
+                return Poll::Ready(None);
+            };
+
+            // Error? just return it and do nothing further.
+            let ev = match ev {
+                Ok(ev) => ev,
+                Err(e) => {
+                    return Poll::Ready(Some(Err(e)));
+                }
+            };
+
+            // React to any actual FollowEvent we get back.
+            let ev = match ev {
+                FollowStreamMsg::Ready(subscription_id) => {
+                    // update the subscription ID we'll use to unpin things.
+                    this.subscription_id = Some(subscription_id.clone().into());
+
+                    FollowStreamMsg::Ready(subscription_id)
+                }
+                FollowStreamMsg::Event(FollowEvent::Initialized(details)) => {
+                    // The first finalized block gets the starting block_num.
+                    let rel_block_num = this.rel_block_num;
+                    let block_ref = this.pin_block_at(rel_block_num, details.finalized_block_hash);
+
+                    FollowStreamMsg::Event(FollowEvent::Initialized(Initialized {
+                        finalized_block_hash: block_ref,
+                        finalized_block_runtime: details.finalized_block_runtime,
+                    }))
+                }
+                FollowStreamMsg::Event(FollowEvent::NewBlock(details)) => {
+                    // One bigger than our parent, and if no parent seen (maybe it was
+                    // unpinned already), then one bigger than the last finalized block num
+                    // as a best guess.
+                    let parent_rel_block_num = this
+                        .pinned
+                        .get(&details.parent_block_hash)
+                        .map(|p| p.rel_block_num)
+                        .unwrap_or(this.rel_block_num);
```

Comment:
This might be a good place to catch rpc-v2 implementation inconsistencies in the wild, but realize that we might not want to polute users with this.
What would you think about having a `log::warning` about the rpc not reporting parent hashes that we expect to exist?

Reply:
I like this idea, buttt if the parent block is unpinned already then it will not exist when we look it up here, so I don't think I can so easily add a log that won't lead to spurious warnings